plugins/woocommerce-blocks/assets/js/base/stores/cart-items.ts

```diff
@@ -0,0 +1,136 @@
+/**
+ * External dependencies
+ */
+import { store } from '@woocommerce/interactivity';
+import type { store as StoreType } from '@wordpress/interactivity'; // Todo: remove once we import from `@wordpress/interactivity`.
+
+type Item = {
+	key?: string;
+	id: number;
+	quantity: number;
+};
+
+type ServerState = {
+	state: {
+		restUrl: string;
+		nonce: string;
+		cart: {
+			items: Item[];
+		};
+	};
+};
+
+export type Store = ServerState & typeof woo;
+
+type StoreAPIError = { message: string; code: string };
+type CartItemResponse = Item | StoreAPIError;
+type CartItemsResponse = Item[] | StoreAPIError;
+
+function isSuccessfulResponse(
+	res: Response,
+	json: CartItemsResponse | CartItemResponse
+): json is Item | Item[] {
+	return res.status.toString().startsWith( '2' );
+}
+
+function generateError( json: StoreAPIError ) {
+	return Object.assign( new Error( json.message || 'Unknown error.' ), {
+		code: json.code || 'unknown_error',
+	} );
+}
+
+let pendingRefresh = false;
+let refreshTimeout = 3000;
+
+// Todo: Remove the type cast once we import from `@wordpress/interactivity`.
+// Question: disable "used before defined" lint rule?
+export const { state, actions } = ( store as typeof StoreType )< Store >(
+	'woocommerce',
+	{} // Todo: Fix this type problem, we should not require an empty object here.
+);
```

Comment:
I am testing to infer the types of the client's store part, but I am not very clear on what is best here, especially because it is not very clean due to the rule that does not allow referencing variables that have not yet been defined.

Reply:
I have reverted this for now and went back to manually typing the whole store.